src/tools/crossgen2/Common/Compiler/DependencyAnalysis/ArrayOfEmbeddedPointersNode.cs

```diff
@@ -97,6 +103,17 @@ public override IEnumerable<DependencyListEntry> GetStaticDependencies(NodeFacto
             }
 
             public override int ClassCode => -66002498;
+
+            public override int CompareToImpl(ISortableNode other, CompilerComparer comparer)
```

Comment:
I'm not sure I understand why we need the interface and why the previous scheme didn't work.

Previously we would call the base `CompareToImpl` implementation in `EmbeddedPointerIndirectionNode` (that we're deleting now). This implementation would call into `CompilerComparer` that would do the thing with `ClassCode` the code is doing here and call `CompareToImpl` as a tie breaker.

How is this different? (Besides removing the `Assert` safeties `CompilerComparer` has.)

Reply:
There was a bug in the `EmbeddedPointerIndirectionNode`. It's possible for its `CompareToImpl` to be called with two different `TTargets`, so the cast to `EmbeddedPointerIndirectionNode<TTarget>` fails. When it fails, we need some way of specifying that the `ISortableNode other` has a `ISortableSymbolNode Target` field that we can then compare. However, you're right that this compare should have been done using CompilerComparer to avoid the code duplication; I'll change that.

I moved the comparison logic to the base classes because @fadimounir and I thought it would be better to keep the comparison logic out of purely abstract classes.

Reply:
I see - this is because `ArrayOfEmbeddedPointersNode` is generic and it doesn't know how to compare members of other instantiations.

I wonder if we could fix this in `EmbeddedPointerIndirectionNode` (and get rid of the variant interface casting in the process) by replacing `interface ISimpleEmbeddedPointerIndirectionNode<out TTarget>` with:

```csharp
interface IIndirectSortableSymbolNode
{
    ISortableSymbolNode Target { get; }
}

public abstract class EmbeddedPointerIndirectionNode<TTarget> : EmbeddedObjectNode, ISortableSymbolNode, IIndirectSortableSymbolNode
    where TTarget : ISortableSymbolNode
{
    // ...

    ISortableSymbolNode IIndirectSortableSymbolNode.Target { get { _targetNode; } }

    // ...

    public override int CompareToImpl(ISortableNode other, CompilerComparer comparer)
    {
        return comparer.Compare(_targetNode, ((IIndirectSortableSymbolNode)other).Target);
    }
}
```

This would encapsulate sorting in one place and the other children of EmbeddedPointerIndirectionNode don't need to override.

Reply:
I would still need to have a way of getting CallSite in the RvaEmbeddedPointerIndirectionNode though (and in the future we might add more fields to be compared in the subclasses).

Reply:
With my recent change, we don't need the CallSite any more because that's being compared in Import.cs now.